lightning/src/offers/mod.rs

```diff
@@ -0,0 +1,408 @@
+// This file is Copyright its original authors, visible in version control
+// history.
+//
+// This file is licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your option.
+// You may not use this file except in accordance with one or both of these
+// licenses.
+
+//! Implementation of Lightning Offers
+//! ([BOLT 12](https://github.com/lightning/bolts/blob/master/12-offer-encoding.md)).
+
+use bitcoin::bech32;
+use bitcoin::bech32::FromBase32;
+use bitcoin::hash_types::BlockHash;
+use bitcoin::secp256k1::{PublicKey, XOnlyPublicKey};
+use bitcoin::secp256k1::schnorr::Signature;
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::time::Duration;
+use ln::PaymentHash;
+use ln::features::OfferFeatures;
+use ln::msgs::DecodeError;
+use util::ser::{HighZeroBytesDroppedVarInt, Readable, WithLength, WithoutLength};
+
+use prelude::*;
+
+#[cfg(feature = "std")]
+use std::time::SystemTime;
+
+///
+#[derive(Debug)]
+pub struct Offer {
+	tlv_stream: OfferTlvStream,
+}
+
+impl Offer {
+	///
+	pub fn description(&self) -> &String {
+		&self.tlv_stream.description.as_ref().unwrap().0
```

Comment:
> Its exactly this kind of thing that I find really gross. If there's a field that's required, it should be required, not an `Option`. It pushes the complexity into the implementation instead of the struct.

Agreed. Converting the TLV stream instead of wrapping would avoid this and is preferable.

> I'm further not really sure why we can't just make it `required` in the TlvStream - why does it have to be an `option` and then unwraped?

With the converting approach, I would prefer to keep it optional in the TLV stream (but required in `Offer`) so as to not mix error types across layers. Otherwise, missing a description would be a `DecodeError` while missing both `node_id` and `paths` would be a `SemanticError`.

Reply:
> Right, but there you're trying to work around limitations in the serialization framework by adding complexity further up...instead of improving the serialization framework to do what you want. If there's something you want from the serialization framework, lets add it, not work around it.

Hmmm... I'm not trying to get around any limitations of the serialization framework nor do I want anything more from it. It's working exactly as I expect it should: decoding a stream of TLV records as defined by each record's type. I don't see a compelling reason why decoding should include semantic checks. This is a standard separation of concerns.

Are you arguing that _all_ semantic checks should be handled by the serialization framework? If not, what is the limiting principle as to what should be included there? It doesn't take long before business logic finds it's way in (e.g., `quantity_min <= quantity_max` -- assuming both are set -- or `signature` is valid for the offer given `node_id` or offer has not expired -- assuming `std` ). There are more complex semantics across `amount`, `currency`, and `chains`, which I haven't touched on yet. Excluding all those, supporting even simple semantics would add considerable complexity and maintenance burden to the serialization macros (e.g., if field A is set must not set field B, field A must only be set of field B is set, etc.).